### PR TITLE
[TS] LPS- 204348 Liferay 7.4 doesnt do a 404 redirect when navigating to errors/code.jsp

### PR DESCRIPTION
**Pull request by IstvanD**
Dear Team,

Could you please review this pull request?

Motivation
=======
In 7.0, when a user navigates to http://localhost:8080/errors/code.jsp, Liferay will display a warning message in the console:

```
2023-12-12 18:30:51.262 WARN  [http-nio-8080-exec-10][code_jsp:157] {code="0", msg="null", uri=null}
```

and redirect to the last valid page.

However, in 7.4, Liferay redirects to a blank page and throws an exception:

```
2023-12-12 18:30:51.273 ERROR [http-nio-8080-exec-10][TextHTMLPortalErrorCodeJSPDynamicInclude:40] Unable to include JSP /dynamic_include/text_html.jsp
org.apache.jasper.JasperException: java.lang.NullPointerException
...
Caused by: java.lang.NullPointerException
	at javax.servlet.jsp.PageContext.getErrorData(PageContext.java:556) ~[org.apache.tomcat-jsp-api.jar:2.3.FR]
	at com.liferay.taglib.servlet.PageContextWrapper.getErrorData(PageContextWrapper.java:88) ~[util-taglib.jar:?]
	at org.apache.jsp.dynamic_005finclude.text_005fhtml_jsp._jspService(text_005fhtml_jsp.java:72) ~[?:?]
```
	
The problem is the `PageContext` class in `javax.servlet.jsp:javax.servlet.jsp-api:2.3.1`

```
public ErrorData getErrorData() {
	return new ErrorData( 
	    (Throwable)getRequest().getAttribute( "javax.servlet.error.exception" ),
	    ((Integer)getRequest().getAttribute( 
		"javax.servlet.error.status_code" )).intValue(),
	    (String)getRequest().getAttribute( "javax.servlet.error.request_uri" ),
	    (String)getRequest().getAttribute( "javax.servlet.error.servlet_name" ) );
}
```

This method assumes the attribute “javax.servler.error.status_code” exists and calls `intValue()` on it.

However, Tomcat’s own `PageContext` in `tomcat-jsp-api:9.0.83` is different:

```
        int status = 0;
        Integer status_code = (Integer) getRequest().getAttribute(RequestDispatcher.ERROR_STATUS_CODE);
        // Avoid NPE if attribute is not set
        if (status_code != null) {
            status = status_code.intValue();
        }
```

It checks if the status_code exists.

The code
```
int code = errorData.getStatusCode();
```
is executed in two places:
1. `code.jsp` which is in `portal-web`
2. `text_html.jsp` which is in the `portal-error-code` module
The `NullPointerException` is thrown from `text_html.jsp` but not from `code.jsp`.
The reason for this is that `portal-web` refers to the vendor-specific tomcat version of `jsp-api` while the `portal-error-code` module refers to the `javax` one.

Proposed Solution
========
In `getErrorData()` instead of returning `super.getErrorData()` a code similar to the one in `tomcat-jsp-api` should be executed.

Steps to Verify
========
Please see the linked LPS for the reproduction steps.
https://liferay.atlassian.net/browse/LPS-204348

Thank you and best regards,
István

### CODE REVIEW

**Review by dantewang — APPROVED** [review]

**Comment by dantewang** [comment]
ci:test:relevant

**Comment by dantewang** [comment]
@IstvanD Code change looks good, could you provide either an integration test or a functional test just in case we break it in the future?

**Review by dantewang — CHANGES_REQUESTED** [review]
Change status

**Comment by IstvanD** [comment]
Hi @dantewang,

I couldn't simulate the use case with an integration test. Do you know what team is responsible for creating functional tests and how to ask them?

Thank you!

**Comment by yuhai** [comment]
Hi @IstvanD, we may create the testIntegration test for the fix. For example,
```
@Test
	public void testAccessErrorsCodeJsp() throws Exception {
		URLUtil.toString(new URL("http://localhost:8080/errors/code.jsp"));
	}
```
`URLUtil.toString(url)` will invoke the current fix. After the fix, there is WARN log `2023-12-21 10:22:59.513 WARN  [http-nio-8080-exec-5][code_jsp:157] {code="0", msg="null", uri=null}` output from the portal side. CI test might think this is failed when seeing WARN log. So you may use `LoggerTestUtil.configureLog4JLogger` to do the assert.

I don't find the test class which includes the test yet. You may create one new test class or find one suited existing test class to place. If you also can't find the test class, I will try to find this place next Tuesday (I will be on PTO).

**Comment by IstvanD** [comment]
Hi @yuhai many thanks for the help! Indeed, I couldn't find a suitable test class, but not even a module, so I created a new one. Please see my force-pushed changes.

**Comment by tinatian** [comment]
ci:test:relevant

**Comment by IstvanD** [comment]
ci:test:relevant

File: modules/apps/portal/portal-error-code-test/src/testIntegration/java/com/liferay/portal/error/code/test/PortalErrorCodeTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new LiferayIntegrationTestRule();
+
+	@Test
+	public void testAccessErrorsCodeJsp() throws Exception {
+		URLUtil.toString(new URL("http://localhost:8080/errors/code.jsp"));
+	}
+
+}
```
Comment (yuhai) [inline modules/apps/portal/portal-error-code-test/src/testIntegration/java/com/liferay/portal/error/code/test/PortalErrorCodeTest.java:36]: @IstvanD, the current test doesn't exist Assert. After the test finished, the portal existed the WARN log `2023-12-27 08:46:05.747 WARN  [http-nio-8080-exec-6][code_jsp:157] {code="0", msg="null", uri=null}`,  I remember ci test will fail the test if it found WARN log existed in the log. So please use `LoggerTestUtil.configureLog4JLogger` to do the Assert. The `code_jsp` loggerName actually is `portal_web.docroot.errors.code_jsp`, so you may use the below codes to assert.
```
try (LogCapture logCapture = LoggerTestUtil.configureLog4JLogger("portal_web.docroot.errors.code_jsp", LoggerTestUtil.WARN)) {
			URLUtil.toString(new URL("http://localhost:8080/errors/code.jsp"));

			List<LogEntry> logEntries = logCapture.getLogEntries();

			Assert.assertEquals(logEntries.toString(), 1, logEntries.size());

			LogEntry logEntry = logEntries.get(0);

			Assert.assertEquals(LoggerTestUtil.WARN, logEntry.getPriority());
			
			String message = logEntry.getMessage();

			Assert.assertEquals(
				"{code=\"0\", msg=\"null\", uri=null}", logEntry.getMessage());
		}
```

**Comment by yuhai** [comment]
ci:stop

**Comment by yuhai** [comment]
@IstvanD, please see the above comments and resend it. 

Thanks,
Hai

**Comment by IstvanD** [comment]
Hi @yuhai 
I have sent the pull with the requested changes:
https://github.com/liferay-core-infra/liferay-portal/pull/6372

Regards,
István

**Pull request closed without merging**